server/src/main/java/com/google/sps/utils/AdRowProcessor.java

```diff
@@ -0,0 +1,118 @@
+/* 
+ * Description: File for building ads and adding them to Firestore
+ * Author: Kira Toal
+ * Date: June 24, 2020
+ */ 
+package com.google.sps.utils;
+
+import com.google.api.core.ApiFuture;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.Firestore;
+import com.google.cloud.firestore.WriteResult;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.FirebaseOptions;
+import com.google.firebase.cloud.FirestoreClient;
+import com.google.sps.utils.Ad;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.time.LocalDate;
+import java.util.Arrays;
+import java.util.ArrayList; 
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class AdRowProcessor {
+
+  private String[] row; 
+
+  public AdRowProcessor(String[] csvRow) {
+    this.row = csvRow; 
+  }
+
+  public Ad createAd() {
+    Ad ad = new Ad.AdBuilder()
+      .id(row[0])
+      .advertiser(row[1])
+      .startDate(row[2]) 
+      .endDate(row[3])
+      .impressionsMin(getImpressionsMin(row[4]))
+      .impressionsMax(getImpressionsMax(row[4]))
+      .ageTargetingEnabled(isAgeTargetingEnabled(row[5]))
+      .genderTarget(getList(row[6]))
+      .geoTarget(getList(row[7]))
+      .spendMin(getLong(row[8]))
+      .spendMax(getLong(row[9]))
+      .headline(row[10].trim())
+      .link(row[11].substring(3)) // trim "Ad" from "Ad {URL}"
+      .content(row[12].trim())
+      .headlineSentiment(row[13].trim())
+      .headlineTerms(row[14].trim())
+      .contentSentiment(row[15].trim())
+      .contentTerms(row[16].trim())
+      .build();
+    System.out.println(ad.toString());
+    return ad;
+  }
+
+  public void addAdToDatabase(Ad ad, int rowIndex, String COLLECTION) throws Exception {
+    FileInputStream serviceAccount = new FileInputStream("../resources/accountKey.json");
```

Comment:
Extract the path `../resources/accountKey.json` to a constant and place it at the top of the file - so we can easily change it later - paths are extremely easy to misspell/misplace

Reply:
Got it!